File: DataStructures/Queue/Queue.java
/**
 * @summary: simple implementation of a Queue of integers using Singly-Linked List with access to head and tail
 *           as the underlying DS.
 * @author: Kayhan Dehghani
 **/
public class Queue {
	private static sLinkedList queue = new sLinkedList();

	public void enqueue(int element) {
		queue.addToTail(element);
	}
	
	public int dequeue() throws IllegalAccessException {
		return queue.removeFromHead();
	}

	public boolean isEmpty() {
		return queue.isEmpty();
	}

	public int getSize() {
		return queue.getSize();
	}

	// returns the next element waiting to be dequeued	
	public int peek() throws IllegalAccessException {
		return queue.getHead(); 
	}
	
	// prints the elements where the most recently added element is to left of the list
	public String toString() {
		return queue.toString();
	}
	
	static class sLinkedList {
		private static Header header = new Header();
		private static int size;

		boolean isEmpty() {
			return header.getHead() == null && header.getTail() == null && size == 0;
		}

		int getSize() {
			return size;
		}

		void addToTail(int element) {
			IntNode newNode = new IntNode(element, null);
			if(this.isEmpty()) {
				header.setHead(newNode);
			} else {
				header.getTail().setNext(newNode); 
			}
			header.setTail(newNode);
			size++;
		}

		int removeFromHead() throws IllegalAccessException {
			if(!this.isEmpty()) {
				IntNode oldHead = header.getHead();
				if(oldHead.getNext() == null) { // stack has one element
					header.setTail(null);    
				} 
				header.setHead(oldHead.getNext());
				size--;
				return oldHead.getVal();
			} else {
				throw new IllegalAccessException("Empty queue!");
			}
		}

		int getHead() throws IllegalAccessException {
			if(!this.isEmpty()) {
				return header.getHead().getVal();
			} else {
				throw new IllegalAccessException("Empty queue!");
			}
		}

		public String toString() {
			String string = "[ ]";
			if(!this.isEmpty()) {
				IntNode currentNode = queue.header.getHead();
				string = "[ " + currentNode.getVal();
				currentNode = currentNode.getNext();
				while(currentNode != null) {
					string += String.format(", " + currentNode.getVal());
					currentNode = currentNode.getNext();
				}
				string += " ]";
			}
			return string;
		}

		class IntNode {
			private int value;
			private IntNode next;
			
			IntNode(int value, IntNode next) {
				this.value = value;
				this.next = next;
			}

			void setNext(IntNode next) {
				this.next = next;
			}

			IntNode getNext() {
				return next;
			}

			int getVal() {
				return value;
			}
		}

		static class Header {
			private IntNode head;
			private IntNode tail;

			void setTail(IntNode tail) {
				this.tail = tail;
			}

			void setHead(IntNode head) {
				this.head = head;
			}

			IntNode getTail() {
				return tail;
			}

			IntNode getHead() {
				return head;
			}
		}
	}
}

File: DataStructures/Queue/Queue.js
/**
* @summary: simple implementation of a Queue(FIFO) using two stacks as underlying data structure.
* @author: Kayhan Dehghani
**/
let process = require('process');


// constructor
Queue = function() {
	this.pushStack = []; // stack on which enqueue() is performed
	this.popStack = []; // stack on which dequeue() is performed
};


Queue.prototype.enqueue = function(obj) {
	this.pushStack.push(obj);
};

Queue.prototype.dequeue = function() {
	/* 
	* if there is nothing to pop from popStack copy everything from pushStack
	* to popoStack in reverse order and empty pushStack.
	*/
	if(this.popStack.length == 0) {
		this.popStack = this.pushStack.reverse();
		this.pushStack = [];
	}
	return this.popStack.pop();
};

Queue.prototype.isEmpty = function() {
	return this.pushStack.length == 0 && this.popStack.length == 0;
};

Queue.prototype.clear = function() {
	this.popStack = [];
	this.pushStack = [];
};

Queue.prototype.print = function() {
	process.stdout.write("{ ");
	this.popStack.forEach(function(element) {
		process.stdout.write(`(${element}) `);
	});
	this.pushStack.forEach(function(element) {
		process.stdout.write(`(${element}) `);
	});
	process.stdout.write(" }\n");
};